docs/design-proposals/multiple-statefulsets.md

```diff
@@ -0,0 +1,417 @@
+
+The `tikv` section is the main tikv statefulset. `tikvs` is the extra tikv statefulsets. They must include all the attributes.
+
+And each statefulset has a separate ConfigMap named by the statefulset name, the `tikv` statefulset name is: `<clusterName>-tikv`, the `tikvs` statefulset name are `<clusterName>-tikv-name-1` and `<clusterName>-tikv-name-2`.
+
+We must add a name attribute(for example: name-1 or name-2) to the `tikvs` section to distinguish the different statefulsets.
+
+The `tikvs` section can be empty.
+
+The same as TidbCluster object, from Figure 3 to Figure 4:
+
+Figure 3:
+
+``` go
+type TidbClusterSpec struct {
+  TiKV            TiKVSpec            `json:"tikv,omitempty"`
+  …
+}
+
+type TiKVStatus struct {
+  StatefulSet     *apps.StatefulSetStatus     `json:"statefulSet,omitempty"`
+  …
+}
+```
+
+Figure 4:
+
+``` go
+type TidbClusterSpec struct {
+  TiKV         TiKVSpec         `json:"tikv,omitempty"`
+  TiKVs        []TiKVSpec       `json:"tikvs,omitempty"`
+  …
+}
+
+type TiKVSpec struct {
+  Name            string           `json:"name"`
+  Replicas        int32            `json:"replicas"`
+  …
+}
+
+type TiKVStatus struct {
+  StatefulSet     *apps.StatefulSetStatus     `json:"statefulSet,omitempty"`
+  StatefulSets    []apps.StatefulSetStatus    `json:"statefulSets,omitempty"`
+  …
+}
+```
+
+### Create:
+
+The TiDB Operator should maintain multiple statefulset objects in the kube apiserver. The startup scripts are the same as before basically. An exception is PD startup script: the bootstrapping Pod may be the main statefulset first Pod.
+
+The TiDB Operator will create extra Services together with multiple statefulsets:
+
+* For PD: no change
+* For TiKV: no change
+* For TiDB: N+1 Services should be created, N client services and 1 main client service
+
+### Upgrade:
+
+With multiple statefulsets supported, the user can modify the spec in the `values.yaml` then the Operator will range over all the statefulsets and select only one statefulset to upgrade at a time. So the basic progress is the same with older design.
+
+The user can modify all the statefulsets, but we only support upgrade one Pod at a time.
+
+For example, the user can change the images from v2.1.0(Figure 5) to v2.2.0(Figure 6) to upgrade the cluster from v2.1.0 to v2.2.0:
+
+Figure 5:
+
+``` yaml
+tikv:
+  replicas: 3
+  image: pingcap/tikv:v2.1.0
+  …
+
+tikvs:
+  - name: name-1
+    replicas: 2
+    image: pingcap/tikv:v2.1.0
+    …
+  - name: name-2
+    replicas: 2
+    image: pingcap/tikv:v2.1.0
+    …
+```
+
+Figure 6:
+
+``` yaml
+tikv:
+  replicas: 3
+  image: pingcap/tikv:v2.2.0
+  …
+
+tikvs:
+  - name: name-1
+    replicas: 2
+    image: pingcap/tikv:v2.2.0
+    …
+  - name: name-2
+    replicas: 2
+    image: pingcap/tikv:v2.2.0
+    …
+```
+
+The operator does not guarantee the order of multiple statefulsets’ upgrade. If the user care about the order indeed. Just: modify one statefulset, wait it done, and then modify the other one.
+
```

Comment:
I think we can have a default upgrade order, such as we can upgrade the main statefulset first, and then upgrade the statefulset in the order of statefulset slice?

Reply:
Done.